### PR TITLE
So long poly2tri, I'd say thanks for all the fish but... 

### PR DESCRIPTION
**Pull request by nirvn**
This PR moves our rubber band and {line,polygon} QML rendering away from OpenGL primitives in favor of a QML Shape item solution. The move has many, many advantages and from the testing I've done comes at *no* tangible speed cost.

The biggest win is stability. We've known for a while now (thanks to sentry) that the poly2tri-driven tessellation of polygons required to display the OpenGL primitives was prone to cause crashes. We were able to nail a few of those crashes by doing some geometry buffering and duplicate vertices removal, but it turned out to never be quite enough.

The release of QField 3.0 seems to also have led to an increase in poly2tri-related crashes.

So, with this change in, we say bye bye to those crashes *for good* :partying_face: 

The second big win has to do with line width. The Qt class we used to display lines - QSGGeometry - states that the line width property is not guaranteed to work on all 3D renderers. And while that did not appear to be an issue under Qt 5, it certainly is with Qt 6. Long story short, several platforms - windows and iOS being ones we care about - have moved to a Direct3D and Metal 3D renderer which don't support line width.

In practice, it meant that on iOS and Windows, people would get excessively narrow - read unusable - rubber band lines when digitizing features or using the measuring tool.

This PR also takes care of that.

While this PR offers visual parity with the old implementation (well, the anti-aliasing looks much better now actually :smile: ), shapes also come with other benefits: we can have round join style, round cap endings, we can use layer effects (i.e. drop shadow), dashed lines, etc.

So long story short: we've closed an important stability issue, and we've unlocked a large amount of styling improvements we might want to rely on in the future.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for cff2139beb08b72d5470df11a4fc46cbe09281f0: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4789-fba7174dbb3323637cd3dcfc167a20765de47b00-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4789-fba7174dbb3323637cd3dcfc167a20765de47b00-arm-neon-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4789-fba7174dbb3323637cd3dcfc167a20765de47b00-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-4789-fba7174dbb3323637cd3dcfc167a20765de47b00-x86-android.apk)
- [**all access arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4789-fba7174dbb3323637cd3dcfc167a20765de47b00-arm64-android.apk)
- [all access arm-neon-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4789-fba7174dbb3323637cd3dcfc167a20765de47b00-arm-neon-android.apk)
- [all access x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4789-fba7174dbb3323637cd3dcfc167a20765de47b00-x64-android.apk)
- [all access x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-4789-fba7174dbb3323637cd3dcfc167a20765de47b00-x86-android.apk)
</details>

File: src/app/main.cpp [hunk 1]
```diff
@@ -16,6 +16,7 @@
  *                                                                         *
  ***************************************************************************/
 
+#include "platformutilities.h"
```
Comment (m-kuhn) [inline src/app/main.cpp:19]: Needed?

File: src/core/linepolygonshape.cpp [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  {
+    QgsCoordinateTransform ct( mGeometry->crs(), mMapSettings->destinationCrs(), QgsProject::instance()->transformContext() );
+    geometry.transform( ct );
+
+    geomType = geometry.type();
+    switch ( geomType )
+    {
+      case Qgis::GeometryType::Line:
+      {
+        QgsMultiPolylineXY lines = geometry.isMultipart() ? geometry.asMultiPolyline() : QgsMultiPolylineXY() << geometry.asPolyline();
```
Comment (m-kuhn) [inline src/core/linepolygonshape.cpp:53]:
```suggestion
        const QgsMultiPolylineXY lines = geometry.isMultipart() ? geometry.asMultiPolyline() : QgsMultiPolylineXY() << geometry.asPolyline();
```

File: src/app/main.cpp (hunk 1 above)
Comment (nirvn) [reply src/app/main.cpp:19]: @m-kuhn , the PR includes one commit that removes a few #include from qgismobileapp.h. As a result, that #include was needed in main.cpp. Just a bit of house cleanup really :)

File: src/core/linepolygonshape.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/linepolygonshape.cpp]: @m-kuhn , well spotted, fixed that (as well as its polygon counterpart and the rubberband code)

**Review by m-kuhn — APPROVED** [review]
Excellent work, @nirvn 
This is a very good move in terms of stability and maintainability, it would be great to get this merged soon so we have as much time as possible for testing.

@suricactus do you mind having a look at this also?

File: src/qml/Rubberband.qml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      strokeColor: rubberbandShape.color
+      strokeWidth: rubberbandShape.lineWidth / rubberbandShape.scale
+      strokeStyle: ShapePath.SolidLine
+      fillColor: rubberbandShape.polylinesType === Qgis.GeometryType.Polygon
+                 ? Qt.hsla(strokeColor.hslHue, strokeColor.hslSaturation, strokeColor.hslLightness, 0.25)
+                 : "transparent"
+      joinStyle: ShapePath.RoundJoin
+      capStyle: ShapePath.RoundCap
+
+      PathPolyline { path: rubberbandShape.polylines[0] }
```
Comment (suricactus) [inline src/qml/Rubberband.qml:33]:
```suggestion
      PathPolyline {
        path: rubberbandShape.polylines[0]
      }
```

File: src/qml/LinePolygon.qml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import Theme 1.0
+import org.qfield 1.0
+import org.qgis 1.0
+
+LinePolygonShape {
+  id: linePolygonShape
+
+  onPolylinesChanged: {
+    var pathElements = []
```
Comment (suricactus) [inline src/qml/LinePolygon.qml:12]:
```suggestion
    const pathElements = []
```

We don't change the reference of that var, so it should be `const`.

File: src/qml/Rubberband.qml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    anchors.fill: parent
+    ShapePath {
+      strokeColor: rubberbandShape.outlineColor
+      strokeWidth: rubberbandShape.lineWidth / rubberbandShape.scale + 2
+      strokeStyle: ShapePath.SolidLine
+      fillColor: "transparent"
+      joinStyle: ShapePath.RoundJoin
+      capStyle: ShapePath.RoundCap
+
+      PathPolyline { path: rubberbandShape.polylines[0] }
```
Comment (suricactus) [inline src/qml/Rubberband.qml:21]:
```suggestion
      PathPolyline {
        path: rubberbandShape.polylines[0]
      }
```

File: src/qml/LinePolygon.qml (hunk 4 above)
Comment (suricactus) [reply src/qml/LinePolygon.qml:12]: Do we add semicolons or not? Personally I am fan of semicolons, the codebase is mixed. What do we do with new code like this? No?

File: src/qml/qgismobileapp.qml [hunk 6]
```diff
@@ -630,8 +631,8 @@ ApplicationWindow {
     /** A rubberband for the different geometry editors **/
     Rubberband {
       id: geometryEditorsRubberband
-      lineWidth: 2.5
-      color: '#80000000'
+      lineWidth: 4
+      color: '#96000000'
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:635]: Since these are defaults, would be easy if we use them instead of explicitly setting `lineWidth` in several places?

**Review by suricactus — APPROVED** [review]
Good stuff. Small comments.

We need some kind of QML guidelines for QML until we get a linter that satisfies our needs. It will make life much easier for the future when we apply that linter to not change pretty much whole codebase.
